internal/admission/validator.go

```diff
@@ -229,48 +236,46 @@ func (validator KongHTTPValidator) ValidateCredential(
-	// in the namespace, as such we pull a list of all consumers from the cached
-	// client to determine if the credentials are referenced.
+	// If we know it's a credentials secret, we can ensure its base-level validity.
+	if err := credsvalidation.ValidateCredentials(&secret); err != nil {
+		return false, fmt.Sprintf("%s: %s", ErrTextConsumerCredentialValidationFailed, err), nil
+	}
+
+	// Credentials are validated further for unique key constraints only if they are referenced by a managed consumer
+	// in the namespace, as such we pull a list of all consumers from the cached client to determine
+	// if the credentials are referenced.
 	managedConsumers, err := validator.listManagedConsumers(ctx)
 	if err != nil {
 		return false, ErrTextConsumerUnretrievable, err
 	}
 
-	// verify whether this secret is referenced by any managed consumer
+	// Verify whether this secret is referenced by any managed consumer.
 	managedConsumersWithReferences := listManagedConsumersReferencingCredentialsSecret(secret, managedConsumers)
 	if len(managedConsumersWithReferences) == 0 {
-		// if no managed consumers reference this secret, its considered
-		// unmanaged and we don't validate it unless it becomes referenced
-		// by a managed consumer at a later time.
+		// If no managed consumers reference this secret, its considered unmanaged, and we don't validate it
+		// unless it becomes referenced by a managed consumer at a later time.
 		return true, "", nil
 	}
 
-	// now that we know at least one managed consumer is referencing this
-	// secret we perform the base-level credentials secret validation.
-	if err := credsvalidation.ValidateCredentials(&secret); err != nil {
-		return false, ErrTextConsumerCredentialValidationFailed, err
-	}
-
-	// if base-level validation passes we move on to create an index of
-	// all managed credentials so that we can verify that the updates to
-	// this secret are not in violation of any unique key constraints.
+	// If base-level validation passed and the credential is referenced by a consumer,
+	// we move on to create an index of all managed credentials so that we can verify that
+	// the updates to this secret are not in violation of any unique key constraints.
 	ignoreSecrets := map[string]map[string]struct{}{secret.Namespace: {secret.Name: {}}}
 	credentialsIndex, err := globalValidationIndexForCredentials(ctx, validator.ManagerClient, managedConsumers, ignoreSecrets)
 	if err != nil {
 		return false, ErrTextConsumerCredentialValidationFailed, err
 	}
 
-	// the index is built, now validate that the newly updated secret
+	// The index is built, now validate that the newly updated secret
 	// is not in violation of any constraints.
 	if err := credentialsIndex.ValidateCredentialsForUniqueKeyConstraints(&secret); err != nil {
-		return false, ErrTextConsumerCredentialValidationFailed, err
+		return false, fmt.Sprintf("%s: %s", ErrTextConsumerCredentialValidationFailed, err), nil
```

Comment:
Why not

```suggestion
		return false, fmt.Sprintf("%s: %w", ErrTextConsumerCredentialValidationFailed, err), nil
```
?

Reply:
It's `fmt.Sprintf`, not `fmt.Errorf`. We're building here a message that will be returned in a response.

Reply:
Ah right 🤦